azkaban-executor/azkaban-emr-jobtype/src/main/java/uk/gov/dwp/dataworks/azkaban/jobtype/EmrLauncherJob.java

```diff
@@ -0,0 +1,95 @@
+        super(jobId, sysProps, jobProps, log);
+    }
+
+    @Override
+    public void run() throws Exception {
+        boolean successful = service().launchClusterAndWaitForStepCompletion(dependency());
+        if (!successful) {
+            throw new Exception("Job failed");
+        }
+    }
+
+    @Override
+    public void cancel() throws Exception {
+        service().cancel();
+    }
+
+    private String metadataTableName() {
+        return this.getJobProps().getString(METADATA_TABLE_PARAMETER_NAME, "data_pipeline_metadata");
+    }
+
+    private String exportDate() {
+        return jobProps.getString(EXPORT_DATE_PARAMETER_NAME, new SimpleDateFormat("yyyy-MM-dd").format(new Date()));
+    }
+
+    private String dependency() {
+        return jobProps.getString(JOB_DEPENDENCIES_PARAMETER_NAME);
+    }
+
+    private synchronized CompositeService service() {
+        if (_service == null) {
+            AmazonDynamoDB dynamoDB = ClientUtility.amazonDynamoDb(awsRegion());
+            String dataProduct = jobProps.getString(DATA_PRODUCT_NAME, jobProps.getString(EMR_LAUNCHER_LAMBDA_PARAMETER_NAME));
+            DependencyService dependencyService = new DependencyService(dynamoDB, metadataTableName(), exportDate());
+            dependencyService.setParent(this);
+            StatusService statusService = new StatusService(dynamoDB, dataProduct, metadataTableName());
+            LaunchInvocationService launchInvocationService = new LaunchInvocationService(
+                    ClientUtility.amazonLambda(awsRegion()), jobProps.getString(EMR_LAUNCHER_LAMBDA_PARAMETER_NAME));
+            launchInvocationService.setParent(this);
+            String logGroup = this.getJobProps().getString(AWS_LOG_GROUP_PARAMETER_NAME, "");
+            AmazonElasticMapReduce emr = ClientUtility.amazonElasticMapReduce(awsRegion());
+            AWSLogs logs = ClientUtility.awsLogs(awsRegion());
+            LogService logService = new LogService(emr, logs, logGroup);
+            logService.setParent(this);
+            EmrProgressService emrProgressService = new EmrProgressService(emr, logService);
+            emrProgressService.setParent(this);
+            NotificationService notificationService = new NotificationService(
+                    ClientUtility.amazonSNS(awsRegion()),
+                    this.getJobProps().getString(TOPIC_PARAMETER_NAME, "Monitoring"),
+                    jobProps.getString(EMR_LAUNCHER_LAMBDA_PARAMETER_NAME));
+
+            this._service = new CompositeService(dependencyService, launchInvocationService, emrProgressService,
+                    notificationService, statusService, emr);
+            this._service.setParent(this);
+
+        }
+        return _service;
+    }
+
+    private String awsRegion() {
+        return this.getSysProps().getString("aws.region", "eu-west-2");
+    }
+
+    public final static String JOB_DEPENDENCIES_PARAMETER_NAME = "job.dependencies";
```

Comment:
Not a fan of putting fields at the bottom of classes when all of our other code puts them at the top! However we have no coding guidelines in this areas...

Reply:
Snapshot sender, htme, k2hb, claimant consumer, historic data importer all have them at the bottom so certainly not all our other code has them at the top.